### ISSUE TITLE
Suggestion: Custom styles for printing & exporting

### ISSUE THREAD

**Issue by ftolsson** [lmparppei/Beat#105]
Okay, so this is probably the easiest thing to implement, should you find it useful.

I am currently struggling with getting translation of my scripts done right and reliably, and could really use some easy identification cross-languages to identify what paragraph we're talking about.

So, I added this to the css:

```css
body {
	counter-reset: paragraph;

...etc
```
and then 

```css
p:not(.scene-heading):not(.more):not(.header-top):not(.page-break):not(.page-break-render)::before {
	font-size: 8pt;
	color: grey;
	position: absolute;
	top: 0;
	left: 8px;
	counter-increment: paragraph;
	content: counter(paragraph) " "
}
```

and it works like a charm.

I'm thinking that with a toggle in the print menu, giving the `body` a class of `.lineNumbersOn` or so, the pseudo-element could be renamed to

```css
.lineNumbersOn p:not(.scene-heading):not(.more):not(.header-top):not(.page-break):not(.page-break-render)::before {
```

and bam, there's a little feature added for anyone who needs it. 

For your consideration!

**Comment by lmparppei** [comment]
I'm trying to keep the export/print menu tight and not too complicated, but this is a good idea. I'll have to figure out how to put in all of the new stuff so it doesn't make the dialog too cramped or complicated.

New toggles include:
- print revisions 
- print notes
- print line numbers
- print only revised pages

**Comment by lmparppei** [comment]
Late night idea: One possibility would be to introduce some sort of Print Plugins, which have the ability to add new CSS rules or add a new step to preprocessing.

**Comment by ftolsson** [comment]
That's a *very* interesting idea! 

And I totally understand your thinking as far as not overloading the print dialog -- in fact (although slightly off topic) would it be all bad to have a separate print settings dialog (perhaps as a section of the quick settings?) rather than have your dialog pop up before the OS print dialog every time?

In all fairness, most of the settings in the print dialog are things you normally don't want to change more than once anyway. And with a close-to-identical preview in both yours and the OS's dialog box, they at least always manage to trip me up because of their similarity. You know: is it now that I'm supposed to press `print` to get the next box, or will pressing `print` cause sixty pages to be spat out of my physical printer because I should have pressed the save button instead...?
What I'm trying to say is perhaps the settings could be collected elsewhere, and not a required stepping stone every time we want to print/export a script. Just a thought. 

Anyways, in an advanced section of some print settings, print plugins might become a really interesting and useful idea!

**Comment by lmparppei** [comment]
I'd say it's not great to have printing options at multiple places, especially outside of the printing phase. It's better to have one dialog containing all the options, with a preview of what's the actual effect of the choices. Maybe make the print dialog expand to right side when clicking *Show Advanced Options*.

**Comment by adiffer** [comment]
Modern word processors tend to separate page formatting choices from print options. Whether it's line numbers or mirror page formatting for double-sided printing, the list of things people want that look like page formatting can grow quite long. Splitting them doesn't mean you'd want to support every page formatting whim people have, but it would help draw a line in the sand for what's tolerated on the print dialog. 8)

**Comment by ftolsson** [comment]
On the topic of this -- does Beat have access to the "App-Specific Print Options" dropdown/pane/area/thing in the MacOS print dialog?

So, mockup ahead -- but if it does, wouldn't this be a nice way to bypass the first print dialog?
(Especially if the OS print dialog can actually *default* to the app-specific choice here, the options would still be clearly presented to us and we'd have all the benefits of Beat's own dialog. But without having to press print twice. (Which I just made to sound like an overwhelming effort, but hey. Laziness is more than just my hobby, it's my lifestyle.))

<img width="879" alt="image" src="https://user-images.githubusercontent.com/25481285/155311632-3ed47bc3-b39d-4d9c-ae48-b1ba48e78d53.png">

**Comment by lmparppei** [comment]
Not sure if that is possible (it probably is), but I think it's not too feasible, because I'm using the same dialog for PDF export, and in that case I'd have to maintain two different codebases with the same UI stuff.

> Modern word processors tend to separate page formatting choices from print options. Whether it's line numbers or mirror page formatting for double-sided printing, the list of things people want that look like page formatting can grow quite long. Splitting them doesn't mean you'd want to support every page formatting whim people have, but it would help draw a line in the sand for what's tolerated on the print dialog. 8)

Exactly, and Beat print dialog is focused on the page layout/preprocessing stuff. I'd like to keep that separate from the actual, technical *printing*.

I might have a solution for this cooking up, and there's already some code in place for custom CSS rules etc. which could be toggled on/off.

**Comment by lmparppei** [comment]
<img width="941" alt="image" src="https://user-images.githubusercontent.com/20226704/165114545-ced92433-9848-48c7-b2f4-d3ee381e5dd2.png">

A quick update about this: I'm adding *Custom Styles*, which are files called `.beatStyle`. They can reside inside the sandboxed user folder (as with plugins), or inside the app container. They show up like this, and can be toggled on/off.

**Comment by ftolsson** [comment]
That seriously sounds awesome!

I've been wanting to ask you about perhaps adding a features for novel writing (but decided that perhaps I'd better get my revised FTOutliner out the door first) but the way I see it, 99% of that can be taken care of with export styles. Very very cool indeed!

**Comment by lmparppei** [comment]
Note that this doesn't have any connection to the paginator, and only affects the appearance.

ALTHOUGH, that being said, I *could* make it so that stylesheets could hook up to some of the paginator values. Something like this:

```css
body {
    /* blah blah */
}

paginator {
    action {
        width: 61chr;
    }
}
```

Of course, there would be only a limited set of values you could control but it could be kind of nice anyway. It would let me move away from the hardcoded character values, too.

**Comment by lmparppei** [comment]
Implemented in 1.94.7
